src/connection.rs

```diff
@@ -514,12 +402,66 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
 
                     return Err(ConnectionError::Closed);
                 }
-                Either::Right((frame, _)) => {
-                    if let Some(stream) = self.on_frame(frame.transpose().map_err(Into::into))? {
-                        return Ok(stream);
+                Event::NewStream(stream) => {
+                    return Ok(stream);
+                }
+            }
+        }
+    }
+
+    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<Event>> {
+        loop {
+            self.garbage_collect();
+
+            if self.socket.poll_ready_unpin(cx).is_ready() {
+                if let Some(frame) = self.pending_frames.pop_front() {
+                    self.socket.start_send_unpin(frame)?;
+                    continue;
+                }
+            }
+
+            match self.socket.poll_flush_unpin(cx)? {
+                Poll::Ready(()) => {}
+                Poll::Pending => {}
+            }
+
+            match self.socket.poll_next_unpin(cx) {
```

Comment:
Following the rust-libp2p coding guidelines, I would prefer `control_receiver` and `stream_receiver` to be polled before `socket`. That said, this should probably happen in a follow up pull request. The current change maintains the status quo.

Reply:
This has been resolved in the meantime.